File: src/chapters/nonlinear-equations/content/code/bisection.cpp
#include <iostream>
#include <functional>
#include <cmath>
using namespace std;

// Bisection root of f on [a, b] (requires f(a)*f(b) < 0).
double bisection(function<double(double)> f, double a, double b, double tol = 1e-12, int max_iter = 200) {
    double fa = f(a);
    for (int k = 0; k < max_iter; ++k) {
        double c = (a + b) / 2, fc = f(c);
        if (fc == 0 || (b - a) / 2 < tol) return c;
        if (fa * fc < 0) b = c; else { a = c; fa = fc; }
    }
    return (a + b) / 2;
}

int main() {
    cout.precision(12);
    cout << bisection([](double x) { return x * x - 2; }, 1, 2) << "\n";
}

File: src/chapters/nonlinear-equations/content/code/bisection.rs
// Bisection root of f on a bracket [a, b] (requires f(a)*f(b) < 0).
fn bisection<F: Fn(f64) -> f64>(f: F, mut a: f64, mut b: f64, tol: f64, max_iter: usize) -> f64 {
    let mut fa = f(a);
    for _ in 0..max_iter {
        let c = (a + b) / 2.0;
        let fc = f(c);
        if fc == 0.0 || (b - a) / 2.0 < tol { return c; }
        if fa * fc < 0.0 { b = c; } else { a = c; fa = fc; }
    }
    (a + b) / 2.0
}
fn main() {
    println!("{}", bisection(|x: f64| x * x - 2.0, 1.0, 2.0, 1e-12, 200));
}
